sb3_contrib/common/maskable/buffers.py

```diff
@@ -0,0 +1,85 @@
+from typing import Generator, NamedTuple, Optional
+
+import numpy as np
+import torch as th
+from gym import spaces
+from stable_baselines3.common.buffers import RolloutBuffer
+from stable_baselines3.common.vec_env import VecNormalize
+
+
+# TODO use dataclass when Python 3.6 support is dropped
+class MaskedRolloutBufferSamples(NamedTuple):
+    observations: th.Tensor
+    actions: th.Tensor
+    old_values: th.Tensor
+    old_log_prob: th.Tensor
+    advantages: th.Tensor
+    returns: th.Tensor
+    action_masks: th.Tensor
+
+
+class MaskedRolloutBuffer(RolloutBuffer):
+    def __init__(self, *args, **kwargs):
+        self.action_masks = None
+        super().__init__(*args, **kwargs)
+
+    def reset(self) -> None:
+        # TODO: need to handle different action_space attrs based on shape
+        action_masks = np.array([], dtype=np.float32)
+        if isinstance(self.action_space, spaces.Discrete):
+            action_masks = np.ones((self.buffer_size, self.n_envs, self.action_space.n), dtype=np.float32)
+        elif isinstance(self.action_space, spaces.MultiDiscrete):
+            action_masks = np.ones(
+                (self.buffer_size, self.n_envs, sum(self.action_space.nvec)),
+                dtype=np.float32,
+            )
+        self.action_masks = action_masks
+
+        super().reset()
+
+    def add(self, *args, action_masks: np.ndarray = None, **kwargs) -> None:
+        """
+        :param action_masks: Masks applied to constrain the choice of possible actions.
+        """
+        if action_masks is not None:
+            self.action_masks[self.pos] = np.array(action_masks).copy()
```

Comment:
```suggestion
            self.action_masks[self.pos] = np.array(action_masks).reshape(1, sum(self.action_space.nvec)).copy()
```

Was necessary for me.